vsintegration/src/FSharp.Editor/CommonHelpers.fs

```diff
@@ -0,0 +1,211 @@
+        | FSharpTokenColorKind.InactiveCode -> ClassificationTypeNames.ExcludedCode 
+        | FSharpTokenColorKind.PreprocessorKeyword -> ClassificationTypeNames.PreprocessorKeyword 
+        | FSharpTokenColorKind.Operator -> ClassificationTypeNames.Operator
+        | FSharpTokenColorKind.TypeName  -> ClassificationTypeNames.ClassName
+        | FSharpTokenColorKind.Default 
+        | _ -> ClassificationTypeNames.Text
+
+    let private scanSourceLine
+        (
+            sourceTokenizer: FSharpSourceTokenizer, 
+            textLine: TextLine, 
+            lineContents: string, 
+            lexState: FSharpTokenizerLexState
+        ) : SourceLineData =
+    
+        let colorMap = Array.create textLine.Span.Length ClassificationTypeNames.Text
+        let lineTokenizer = sourceTokenizer.CreateLineTokenizer(lineContents)
+            
+        let scanAndColorNextToken(lineTokenizer: FSharpLineTokenizer, lexState: Ref<FSharpTokenizerLexState>) : Option<FSharpTokenInfo> =
+            let tokenInfoOption, nextLexState = lineTokenizer.ScanToken(lexState.Value)
+            lexState.Value <- nextLexState
+            if tokenInfoOption.IsSome then
+                let classificationType = compilerTokenToRoslynToken(tokenInfoOption.Value.ColorClass)
+                for i = tokenInfoOption.Value.LeftColumn to tokenInfoOption.Value.RightColumn do
+                    Array.set colorMap i classificationType
+            tokenInfoOption
+
+        let previousLexState = ref lexState
+        let mutable tokenInfoOption = scanAndColorNextToken(lineTokenizer, previousLexState)
+        while tokenInfoOption.IsSome do
+            tokenInfoOption <- scanAndColorNextToken(lineTokenizer, previousLexState)
+
+        let mutable startPosition = 0
+        let mutable endPosition = startPosition
+        let classifiedSpans = new List<ClassifiedSpan>()
+
+        while startPosition < colorMap.Length do
+            let classificationType = colorMap.[startPosition]
+            endPosition <- startPosition
+            while endPosition < colorMap.Length && classificationType = colorMap.[endPosition] do
+                endPosition <- endPosition + 1
+            let textSpan = new TextSpan(textLine.Start + startPosition, endPosition - startPosition)
+            classifiedSpans.Add(new ClassifiedSpan(classificationType, textSpan))
+            startPosition <- endPosition
+
+        SourceLineData(textLine.Start, lexState, previousLexState.Value, lineContents.GetHashCode(), classifiedSpans)
+
+    let getColorizationData
+        (
+            documentKey: DocumentId, 
+            sourceText: SourceText, 
+            textSpan: TextSpan, 
+            fileName: string option, 
+            defines: string list, 
+            cancellationToken: CancellationToken
+        ) : List<ClassifiedSpan> =
+        try
+            let sourceTokenizer = tokenizers.GetValue ((defines, fileName), fun _ -> FSharpSourceTokenizer(defines, fileName))
```

Comment:
I don't think this cache works given that CWT referential identity of an object as a key